File: numguess.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>

int main()
{
	int i, num = 0, guess = 0, tries = 0, limit = 0, max_tries = 0;
	char name[100], buffer[100];
	srand(time(NULL));
	
	printf("Welcome to NumGuess C version!\n\n");
	printf("Enter your name: ");
	fgets(name, sizeof(name) - 1, stdin);
	
	// Make sure we have a nice zero terminated string
	name[99] = '\0';
	for(i = 0; i < sizeof(name); i++) {
		if(name[i] == '\n') {
			name[i] = '\0';
			break;
		}
	}
	
	if(name[0] == '\0') strcpy(name, "Player");
	
	printf("\nWelcome %s, enter upper limit: ", name);
	fgets(buffer, sizeof(buffer), stdin);
	limit = atoi(buffer);
	if(limit < 10) limit = 10;
	
	max_tries = (int)floor(log2(limit)) + 1;
	
	while(1) {
		num = rand() % limit + 1;
		guess = 0;
		tries = 0;
		
		printf("\nGuess my number between 1 and %d!\n", limit);
		while(num != guess) {
			printf("\nGuess: ");

			fgets(buffer, sizeof(buffer), stdin);
			guess = atoi(buffer);
			
			if((guess < 1) || (guess > limit)) {
				printf("Out of range.");
			} else {
				tries++;
				if(num < guess) {
					printf("Too high!");
				} else if(num > guess) {
					printf("Too low!");
				}
			}
		}
		printf("\nWell done %s, you guessed my number from %d %s!\n", name, tries, tries == 1 ? "try" : "tries");
		
		if(tries == 1) {
			printf("You're one lucky bastard!");
		} else if(tries < max_tries) {
			printf("You are the master of this game!");
		} else if(tries == max_tries) {
			printf("You are a machine!");
		} else if(tries <= max_tries * 1.1) {
			printf("Very good result!");
		} else if(tries <= limit) {
			printf("Try harder, you can do better!");
		} else {
			printf("I find your lack of skill disturbing!");
		}
		
		printf("\nPlay again [Y/N]? ");
		
		fgets(buffer, sizeof(buffer), stdin);
		if((buffer[0] != 'y') && (buffer[0] != 'Y')) break;
	}
	
	printf("\nOkay, bye.\n");

	return 0;
}

File: numguess.swift
import Foundation

// Reads a string from the terminal with an optional default value
func inputString(prompt: String, def: String = "") -> String {
	print(prompt + " ", terminator: "")
	fflush(__stdoutp)
	let handle = NSFileHandle.fileHandleWithStandardInput()
	let data = handle.availableData
	var result = NSString(data: data, encoding: NSUTF8StringEncoding) as! String
	result = result.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceAndNewlineCharacterSet())
	return result == "" ? def : result
}

// Reads a number from the terminal with an optional default and minimum value
func inputNumber(prompt: String, def: Int? = nil, min: Int? = nil) -> Int? {
	var value = Int(inputString(prompt))
	if value == nil { return def }
	if min != nil { value = value < min ? min : value }
	return value
}

// Seed the random generator
srandom(UInt32(time(nil)))

print("Welcome to NumGuess Swift version!\n")
var name = inputString("Enter your name:", def: "Player")
var limit = inputNumber("\nWelcome \(name), enter upper limit:", def: 10, min: 10)!
let maxTries = Int(log2(Double(limit))) + 1
let maxTries10 = Int(Double(maxTries) * 1.1)

repeat {
	print("\nGuess my number between 1 and \(limit)!\n")
	var num = random() % limit + 1
	var tries = 0
	var guess : Int?

	repeat {
		guess = inputNumber("Guess:")

		switch guess {
		case nil:
			print("That's just plain wrong.")
		case let g? where 1...limit ~= g:
			tries += 1
			print(guess < num ? "Too low!" : "Too high!")
		default:
			print("Out of range.")
		}

	} while guess != num

	let triesWord = tries == 1 ? "try" : "tries"
	print("\nWell done \(name), you guessed my number from \(tries) \(triesWord)!")

	switch tries {
	case 1:                     print("You're one lucky bastard!")
	case 2..<maxTries:          print("You are the master of this game!")
	case maxTries:              print("You are a machine!")
	case maxTries...maxTries10: print("Very good result!")
	case maxTries10...limit:    print("Try harder, you can do better!")
	default:                    print("I find your lack of skill disturbing!")
	}

} while "Y" == inputString("Play again [y/N]?").uppercaseString

print("\nOkay, bye.")
